File: superstar/applies_when.py
"""applies_when — pure-logic predicate evaluator for rule frontmatter.

Lives in the framework-agnostic `superstar/` package because it's reused
by the Django decisioning service AND the standalone smoke scripts. No
Django imports.

## Why this exists

The LLM citation guard catches structurally-invented `rule_id` values
(model cited an ID that wasn't retrieved). But the model can still cite
a *real* rule_id whose `applies_when` conditions don't actually match
the request payload — e.g. citing NSD-LOCK-003 (PU/Membrane only) for a
Laminate request. This verifier closes that gap mechanically.

If any cited rule's conditions fail, the decisioning service forces
escalate. Same shape as the existing guards.

## DSL

A frontmatter `applies_when` block is a mapping of `field -> predicate`.
All conditions must match (implicit AND) for the rule to apply.

Predicates:

| Form                                        | Meaning                                |
|---------------------------------------------|----------------------------------------|
| `field: <scalar>`                           | equality                               |
| `field: [v1, v2]`                           | membership: payload[field] in list     |
| `field: {gte: N}`                           | payload[field] >= N                    |
| `field: {gt: N}`                            | payload[field] > N                     |
| `field: {lte: N}`                           | payload[field] <= N                    |
| `field: {lt: N}`                            | payload[field] < N                     |
| `field: {between: [a, b]}`                  | a <= payload[field] <= b               |
| `field: {not_in: [...]}`                    | payload[field] not in list             |
| `field: {not: <scalar>}`                    | inequality                             |
| `field: {has_any: [...]}`                   | array-valued field intersects list     |

A missing payload field counts as a condition failure unless the
predicate is `{optional: true}` (not currently supported — extend if a
real use case appears).

## Return value

`applies_to(conditions, payload) -> (bool, list[str])`

Returns `(applies, reasons_failed)`. When `applies` is `False`, the list
explains exactly which condition didn't match — usable in audit logs.
"""
from __future__ import annotations

from typing import Any

# Operator names recognized in dict predicates. Anything else is a parse error.
_KNOWN_OPS = frozenset({"gte", "gt", "lte", "lt", "between", "not_in", "not", "has_any"})


def applies_to(conditions: dict | None, payload: dict) -> tuple[bool, list[str]]:
    """Evaluate a rule's `applies_when` block against a request payload.

    If `conditions` is None or empty, the rule is considered to apply
    universally (no constraints declared). Callers can opt to treat that
    differently if they prefer strict matching.
    """
    if not conditions:
        return True, []

    if not isinstance(conditions, dict):
        return False, [f"applies_when must be a mapping, got {type(conditions).__name__}"]

    failures: list[str] = []
    for field, predicate in conditions.items():
        ok, reason = _check(field, payload.get(field), predicate)
        if not ok:
            failures.append(reason)

    return (not failures, failures)


def _check(field: str, value: Any, predicate: Any) -> tuple[bool, str]:
    # Bare scalar → equality.
    if isinstance(predicate, (str, int, float, bool)) and not isinstance(predicate, bool):
        # bool is a subclass of int in Python; handle separately to keep semantics tight
        if value == predicate:
            return True, ""
        return False, f"{field} = {value!r}, expected {predicate!r}"

    # Bare bool predicate (explicit branch — order matters).
    if isinstance(predicate, bool):
        if value == predicate:
            return True, ""
        return False, f"{field} = {value!r}, expected {predicate!r}"

    # List → membership.
    if isinstance(predicate, list):
        if value in predicate:
            return True, ""
        return False, f"{field} = {value!r}, expected one of {predicate}"

    # Dict → operator predicate.
    if isinstance(predicate, dict):
        if len(predicate) != 1:
            return False, f"{field}: predicate dict must have exactly one operator, got {list(predicate)}"
        op, arg = next(iter(predicate.items()))
        if op not in _KNOWN_OPS:
            return False, f"{field}: unknown operator {op!r} (known: {sorted(_KNOWN_OPS)})"
        return _apply_op(field, value, op, arg)

    return False, f"{field}: unsupported predicate type {type(predicate).__name__}: {predicate!r}"


def _apply_op(field: str, value: Any, op: str, arg: Any) -> tuple[bool, str]:
    if op in ("gte", "gt", "lte", "lt"):
        if value is None:
            return False, f"{field}: cannot compare {op} {arg!r} — payload value is missing"
        try:
            v = float(value)
            a = float(arg)
        except (TypeError, ValueError):
            return False, f"{field}: cannot compare non-numeric {value!r} {op} {arg!r}"
        cmp = {"gte": v >= a, "gt": v > a, "lte": v <= a, "lt": v < a}[op]
        if cmp:
            return True, ""
        return False, f"{field} = {value!r}, expected {op} {arg!r}"

    if op == "between":
        if not (isinstance(arg, list) and len(arg) == 2):
            return False, f"{field}: between expects [low, high], got {arg!r}"
        if value is None:
            return False, f"{field}: cannot range-check — payload value is missing"
        try:
            v = float(value)
            lo, hi = float(arg[0]), float(arg[1])
        except (TypeError, ValueError):
            return False, f"{field}: non-numeric values in between check"
        if lo <= v <= hi:
            return True, ""
        return False, f"{field} = {value!r}, expected between {arg[0]} and {arg[1]}"

    if op == "not_in":
        if not isinstance(arg, list):
            return False, f"{field}: not_in expects a list, got {arg!r}"
        if value not in arg:
            return True, ""
        return False, f"{field} = {value!r}, expected NOT one of {arg}"

    if op == "not":
        if value != arg:
            return True, ""
        return False, f"{field} = {value!r}, expected != {arg!r}"

    if op == "has_any":
        if not isinstance(arg, list):
            return False, f"{field}: has_any expects a list, got {arg!r}"
        if isinstance(value, list) and any(v in arg for v in value):
            return True, ""
        return False, f"{field} = {value!r}, expected at least one of {arg}"

    # Unreachable because _check already filtered.
    return False, f"{field}: unhandled operator {op}"

File: frontend/src/lib/appliesWhen.ts
/**
 * Client-side twin of superstar/applies_when.py.
 *
 * Same DSL, same operators, same semantics. Used by DynamicForm to evaluate
 * show_if / choices_if conditions in real time as the user types.
 *
 * Operators:
 *   bare scalar              equality
 *   list                     membership
 *   { gte | gt | lte | lt }  numeric comparisons
 *   { between: [a, b] }      inclusive numeric range
 *   { not_in: [...] }        not in list
 *   { not: x }               inequality
 *   { has_any: [...] }       array intersection
 *
 * Empty/null conditions → applies = true (no constraints).
 */

export type Predicate = unknown;
export type ConditionsMap = Record<string, Predicate> | null | undefined;

const KNOWN_OPS = new Set([
  "gte", "gt", "lte", "lt", "between", "not_in", "not", "has_any",
]);

export function appliesTo(
  conditions: ConditionsMap,
  payload: Record<string, unknown>,
): { applies: boolean; reasons: string[] } {
  if (!conditions || typeof conditions !== "object" || Object.keys(conditions).length === 0) {
    return { applies: true, reasons: [] };
  }
  const failures: string[] = [];
  for (const [field, predicate] of Object.entries(conditions)) {
    const ok = checkField(field, payload[field], predicate, failures);
    if (!ok) {
      // checkField pushes a reason; loop continues so reasons are exhaustive.
    }
  }
  return { applies: failures.length === 0, reasons: failures };
}

function checkField(
  field: string,
  value: unknown,
  predicate: Predicate,
  failures: string[],
): boolean {
  // Bare scalar → equality
  if (
    typeof predicate === "string" ||
    typeof predicate === "number" ||
    typeof predicate === "boolean"
  ) {
    if (value === predicate) return true;
    failures.push(`${field} = ${JSON.stringify(value)}, expected ${JSON.stringify(predicate)}`);
    return false;
  }

  // List → membership
  if (Array.isArray(predicate)) {
    if (predicate.some((p) => p === value)) return true;
    failures.push(`${field} = ${JSON.stringify(value)}, expected one of ${JSON.stringify(predicate)}`);
    return false;
  }

  // Dict → operator
  if (predicate && typeof predicate === "object") {
    const entries = Object.entries(predicate as Record<string, unknown>);
    if (entries.length !== 1) {
      failures.push(`${field}: predicate must have exactly one operator, got ${entries.length}`);
      return false;
    }
    const [op, arg] = entries[0];
    if (!KNOWN_OPS.has(op)) {
      failures.push(`${field}: unknown operator ${JSON.stringify(op)}`);
      return false;
    }
    return applyOp(field, value, op, arg, failures);
  }

  failures.push(`${field}: unsupported predicate ${JSON.stringify(predicate)}`);
  return false;
}

function applyOp(
  field: string,
  value: unknown,
  op: string,
  arg: unknown,
  failures: string[],
): boolean {
  if (op === "not") {
    if (value !== arg) return true;
    failures.push(`${field}: not = ${JSON.stringify(arg)} but value matched`);
    return false;
  }

  if (op === "not_in") {
    if (!Array.isArray(arg)) {
      failures.push(`${field}: not_in requires a list`);
      return false;
    }
    if (!arg.some((a) => a === value)) return true;
    failures.push(`${field} = ${JSON.stringify(value)} but is in ${JSON.stringify(arg)}`);
    return false;
  }

  if (op === "has_any") {
    if (!Array.isArray(arg)) {
      failures.push(`${field}: has_any requires a list`);
      return false;
    }
    if (Array.isArray(value) && value.some((v) => arg.includes(v))) return true;
    failures.push(`${field} has none of ${JSON.stringify(arg)}`);
    return false;
  }

  // Numeric ops
  const numericOps = ["gte", "gt", "lte", "lt"];
  if (numericOps.includes(op)) {
    if (value === null || value === undefined || value === "") {
      failures.push(`${field}: cannot compare ${op} ${JSON.stringify(arg)} — missing value`);
      return false;
    }
    const v = Number(value);
    const a = Number(arg);
    if (Number.isNaN(v) || Number.isNaN(a)) {
      failures.push(`${field}: non-numeric comparison`);
      return false;
    }
    const pass =
      op === "gte" ? v >= a :
      op === "gt"  ? v >  a :
      op === "lte" ? v <= a :
                     v <  a;
    if (pass) return true;
    failures.push(`${field} = ${v}, expected ${op} ${a}`);
    return false;
  }

  if (op === "between") {
    if (!Array.isArray(arg) || arg.length !== 2) {
      failures.push(`${field}: between requires [low, high]`);
      return false;
    }
    if (value === null || value === undefined || value === "") {
      failures.push(`${field}: missing value for between`);
      return false;
    }
    const v = Number(value);
    const lo = Number(arg[0]);
    const hi = Number(arg[1]);
    if (Number.isNaN(v) || Number.isNaN(lo) || Number.isNaN(hi)) {
      failures.push(`${field}: non-numeric between`);
      return false;
    }
    if (v >= lo && v <= hi) return true;
    failures.push(`${field} = ${v}, expected between ${lo} and ${hi}`);
    return false;
  }

  failures.push(`${field}: unhandled operator ${op}`);
  return false;
}


/**
 * Resolve the active choices for an enum field given the current payload.
 * Mirrors backend _active_choices: first matching rule wins, fall back to
 * the field's static `choices`.
 */
export function activeChoices(
  field: { choices?: string[]; choices_if?: Array<{ conditions: ConditionsMap; choices: string[] }> },
  payload: Record<string, unknown>,
): string[] {
  for (const rule of field.choices_if ?? []) {
    if (appliesTo(rule.conditions, payload).applies) {
      return rule.choices ?? [];
    }
  }
  return field.choices ?? [];
}
